[Bug 16279] Prevent crash when creating notifications with empty payload

livecodepanos opened the pull request.

livecodepanos committed:
[Bug 16279] Prevent crash when calling mobileCreateLocalNotification

and payload is empty.

livecodepanos committed:
[Bug 16279] Added bugfix note

runrevmark commented:
@livecode-vulcan review ok f6caacc

livecode-vulcan commented:
💙 review by @runrevmark ok f6caacccf4a435b2eb655e11fdfbb5badd854767

livecode-vulcan commented:
😎 test success f6caacccf4a435b2eb655e11fdfbb5badd854767
- try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/952)
- try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/960)
- try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/949)
- try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/953)
- try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/962)
- try-community-win-x86: [success](https://vulcan.livecode.com/builders/try-community-win-x86/builds/971)
- try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/966)

livecodepanos committed:
[Bug 16279] Prevent crash when sending remote notifications with empty payload

If the payload of a *remote* notification is empty, JSON null will be deserialized to NSNull.
Sending a message to NSNull will crash the app, thus make sure we convert NSNull to nil.

runrevmark commented on engine/src/mbliphoneapp.mm:
```diff
@@ -505,7 +520,19 @@ - (void)application:(UIApplication *)p_application didReceiveRemoteNotification:
     UIApplicationState t_state = [p_application applicationState];
     MCAutoStringRef t_mc_push_notification_text;
     NSString *t_reminder_text = [p_dictionary objectForKey:@"payload"];
```
This isn't strictly correct if the return of objectForKey: is NSNull and not NSString. It is better to make this explicit:

```
id t_reminder_text_value = [p_notification.userInfo objectForKey:@"payload"];
NSString *t_reminder_text;
if ([t_reminder_text_value isKindOfClass: [NSString class]])
  t_reminder_text = t_reminder_text_value;
else
  t_reminder_text = nil
```

livecodepanos committed:
[Bug 16279] Better handling of NSNull returned by objectForKey:

runrevmark commented on engine/src/mbliphoneapp.mm:
```diff
@@ -360,8 +364,18 @@ - (BOOL)application:(UIApplication *)p_application didFinishLaunchingWithOptions
+		// Prevent crash when launching the app from the notifications screen:
+		// If the payload is empty, JSON null will be deserialized to NSNull. Unlike nil,
+		// we cannot send (most) messages to NSNull
+		id t_pending_push_notification_value = [[t_push_notification objectForKey:@"payload"] retain];
```
Is there a need for the retain here?

livecodepanos replied:
I guess yes, because of bug 10391:
http://quality.runrev.com/show_bug.cgi?id=10391

runrevmark replied:
Ah sorry - I see - don't retain here - retain when you assign it to m_pending_push_notification :)

livecodepanos replied:
Oops my bad :)

livecodepanos committed:
[Bug 16279] Retain notification in the correct place

runrevmark commented:
@livecode-vulcan review ok 419ef79

livecode-vulcan commented:
💙 review by @runrevmark ok 419ef79d655d9e9f71f58e1cdf93110825abb783

livecode-vulcan commented:
😎 test success 419ef79d655d9e9f71f58e1cdf93110825abb783
- try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/1039)
- try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/1047)
- try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/1036)
- try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/1040)
- try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/1050)
- try-community-win-x86: [success](https://vulcan.livecode.com/builders/try-community-win-x86/builds/1057)
- try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/1055)
